Update EIP-8141: note sponsor repayment reversibility and batch the repayment in Example 3 (#11956)
ethereum/EIPs: EIPS/eip-8141.md

| ----- | ------- | ----------- | ------------------ | ------------- | ----- | ---------------------- |
| 0     | VERIFY  | ENTRY_POINT | APPROVE_EXECUTION  | Null (sender) | 0     | Empty                  |
| 1     | VERIFY  | ENTRY_POINT | APPROVE_PAYMENT    | Sponsor       | 0     | Sponsor data           |
| 2     | SENDER  | Sender      | ATOMIC_BATCH_FLAG  | ERC-20        | 0     | transfer(Sponsor,fees) |
| 3     | SENDER  | Sender      | APPROVE_SCOPE_NONE | Target addr   | 0     | Call data              |
| 4     | DEFAULT | ENTRY_POINT | APPROVE_SCOPE_NONE | Sponsor       | 0     | Post op call           |

- Frame 0: Uses the sender's signature entry with empty `msg` and calls `APPROVE(APPROVE_EXECUTION)` to authorize execution from sender.

> - Frame 1: Inspects signature metadata in `tx.signatures` if needed, and checks that the next frame is an ERC-20 send of the right size to the sponsor and that it shares an atomic batch with the sponsee's intended call. Calls `APPROVE(APPROVE_PAYMENT)` to authorize payment.

Reviewer: ```suggestion
- Frame 1: Inspects signature metadata in `tx.signatures` if needed, and checks that the next frame is an ERC-20 send of the right size to the sponsor, that it begins an atomic batch, and that no `SENDER` frame lies outside that batch. Calls `APPROVE(APPROVE_PAYMENT)` to authorize payment.
```

- Frame 2: Sends tokens to sponsor. It sets `ATOMIC_BATCH_FLAG` so it is batched with frame 3: if either the repayment or the intended call reverts, both are unrolled. This prevents the sponsee from letting the repayment revert while its intended call still executes.
- Frame 3: User's intended call, and the terminating frame of the atomic batch begun in frame 2.
- Frame 4 (optional): Check unpaid gas, refund tokens, possibly convert tokens to ETH on an AMM.

> Note: payment collected at `APPROVE(APPROVE_PAYMENT)` in frame 1 is final and is not reverted by a later frame. Because frames 2 and 3 share an atomic batch, a sponsee that zeroes its ERC-20 balance only causes its own intended call to be skipped and gains nothing; the sponsor still pays gas for the reverted attempt, which is the residual, unprofitable griefing risk inherent to public sponsorship.

Reviewer: ```suggestion
Note: payment collected at `APPROVE(APPROVE_PAYMENT)` in frame 1 is final and is not reverted by a later frame. Because every `SENDER` frame is inside the atomic batch begun by the repayment, a sponsee that zeroes its ERC-20 balance only causes its own intended call to be skipped and gains nothing; the sponsor still pays gas for the reverted attempt, which is the residual, unprofitable griefing risk inherent to public sponsorship.
```

`FRAMEPARAM`, `FRAMEDATALOAD`, and `FRAMEDATACOPY` allow validation code to inspect other frames, including later `SENDER` frames and their `value`s. As a result, paymasters and other `VERIFY` frames can observe user operation parameters before approval and may condition their behavior on that information. Users should therefore treat non-`VERIFY` frame parameters and data as visible to validation logic and should not rely on untrusted paymasters or verifiers to keep such information private.

#### Sponsor Repayment Reversibility

> Payment collected when a frame calls `APPROVE(APPROVE_PAYMENT)` or `APPROVE(APPROVE_EXECUTION_AND_PAYMENT)` is not reverted by the failure of any later frame. A frame that repays a sponsor (for example, an ERC-20 transfer to the sponsor) executing after payment approval is an ordinary frame: if it is not part of an atomic batch, its revert discards only that frame while the remaining frames — including the sponsee's intended operation — continue. A sponsee could therefore let the repayment frame revert while its intended operation still executes, obtaining execution without paying the sponsor. Sponsors should place the repayment frame and the sponsee's intended `SENDER` frame(s) in the same atomic batch so that they share fate, as shown in Example 3; a `VERIFY` frame may inspect frame flags via `FRAMEPARAM` to confirm this structure before calling `APPROVE(APPROVE_PAYMENT)`. Only the gas payment itself is guaranteed final; all repayment and other post-payment effects are reversible.

Reviewer: ```suggestion
Payment collected when a frame calls `APPROVE(APPROVE_PAYMENT)` or `APPROVE(APPROVE_EXECUTION_AND_PAYMENT)` is not reverted by the failure of any later frame. A frame that repays a sponsor (for example, an ERC-20 transfer to the sponsor) executing after payment approval is an ordinary frame: if it is not part of an atomic batch, its revert discards only that frame while the remaining frames, including the sponsee's intended operation, continue. A sponsee could therefore let the repayment frame revert while its intended operation still executes, obtaining execution without paying the sponsor. Sponsors should place the repayment frame and the sponsee's intended `SENDER` frame(s) in the same atomic batch so that they share fate, as shown in Example 3. The check must cover the whole frame list: the repayment frame begins an atomic batch and no `SENDER` frame lies outside it. A `SENDER` frame placed before the repayment or after the batch's terminal frame executes regardless of the repayment's fate. A `VERIFY` frame can perform this scan with `FRAMEPARAM` (mode and flags of every frame) before calling `APPROVE(APPROVE_PAYMENT)`. Only the gas payment itself is guaranteed final; all repayment and other post-payment effects are reversible.
```
